stm32/uart: Fix UART timeout issue with low baudrate.

yn386 opened the pull request:
### MicroPython Version
v1.20.0-261-g813d559bc

### Environment
NUCLEO-G474RE

### How to reproduce
Execute this code:
```python
from pyb import UART
uart = UART(1, 4800, bits=8, parity=None, stop=1)
print(uart)

print(uart.any())
print(uart.write("123"))
print(uart.write(b"abcd"))
print(uart.writechar(1))
```

Got result:
```
UART(1, baudrate=4800, bits=8, parity=None, stop=1, flow=0, timeout=0, timeout_char=4, rxbuf=64)
0
3
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
OSError: [Errno 110] ETIMEDOUT
```

### Detail of changes
`timeout_char` is calculated at
https://github.com/micropython/micropython/blob/813d559bc098eeaa1c6e0fa1deff92e666c0b458/ports/stm32/machine_uart.c#L253
However, if `13000 / baudrate` is not divisible by baudrate (such as 4800, 9600, ...), calculated timeout will be less than actual timeout required.
To solve this issue, this PR rounds up timeout_char if `13000 / baudrate` is not divisible by baudrate.

Since this PR affects test results using tests/pyb/uart.py, I also modified expected test result.

### Effect of changes
Be able to write values with low baudlate.
```
UART(1, baudrate=4800, bits=8, parity=None, stop=1, flow=0, timeout=0, timeout_char=5, rxbuf=64)
0
3
4
None
```

yn386 committed:
stm32/uart: Fix UART timeout issue with low baudrate on G4 MCUs.

With using UART FIFO, the timeout should be long enough that FIFO becomes
empty.  Since previous data transfer may be ongoing, the timeout must be
timeout_char multiplied by FIFO size + 1.

Signed-off-by: Yuuki NAGAO <wf.yn386@gmail.com>

dpgeorge commented:
This fix does not look correct.  The `timeout_char` calculation is already rounded-up (by 2).

For 4800 baud, `min_timeout_char` will be calculated as 4, which is 4ms as a timeout.  The time for a 10-bit character to be output on the UART at 4800 baud is `10 / 4800 = 2.08ms`.  So there should be plenty of margin already in the timeout.

yn386 commented:
Thank you for reviewing. I will reconsider why this issue was occurred.

yn386 commented:
I found why timeout is not long enough even rounded up.
This issue has a relationship with using UART FIFO.

For G4 MCUs, UART FIFO is enabled here:
https://github.com/micropython/micropython/blob/7d66ae603d91157069d7cdee065919a61f678fb9/ports/stm32/uart.c#L584-L586
With using UART FIFO, the timeout should be long enough that FIFO becomes empty.
For example, STM32G4 has 8 words FIFO. Since the data is queued, the code
```python
print(uart.write(b"abcd"))
```
with 4800 baudrate needs about `4 * 10 / 4800 = 8.3ms` until `UART_FLAG_TC` became 1.
This is longer than calculated timeout (4ms).
The maximum duration that FIFO becomes empty will be `8 * 10 / 4800 = 16.7ms`.

Since the `timeout_char` is long enough for 1 char, we just multiply with FIFO size.

The boards those does not have or use UART FIFO (such as F4 MCUs) does not occur this issue.
However, this implementation will be needed when the UART FIFO makes to be enabled with having UART FIFO boards.

Now I rebased and fixed this issue (f093dfd7).

yn386 commented:
I noticed that the timeout should consider previous data transfer may be ongoing.
I modified previous [comment](https://github.com/micropython/micropython/pull/11915#issuecomment-1636969428) and fixed it (d1d1ab93).

dpgeorge commented:
Thanks for updating, now merged.
